flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -21,519 +21,897 @@
-		SlotRequest chosenRequest = chooseRequestToFulfill(freeSlot, pendingSlotRequests);
+	private void internalRequestSlot(PendingSlotRequest pendingSlotRequest) throws ResourceManagerException {
+		TaskManagerSlot taskManagerSlot = findMatchingSlot(pendingSlotRequest.getResourceProfile());
 
-		if (chosenRequest != null) {
-			final AllocationID allocationId = chosenRequest.getAllocationId();
-			final SlotRequest slotRequest = pendingSlotRequests.remove(allocationId);
+		if (taskManagerSlot != null) {
+			allocateSlot(taskManagerSlot, pendingSlotRequest);
+		} else {
+			final UUID timeoutIdentifier = UUID.randomUUID();
+			final AllocationID allocationId = pendingSlotRequest.getAllocationId();
+
+			// register timeout for slot request
+			ScheduledFuture<?> timeoutFuture = scheduledExecutor.schedule(new Runnable() {
+				@Override
+				public void run() {
+					mainThreadExecutor.execute(new Runnable() {
+						@Override
+						public void run() {
+							timeoutSlotRequest(allocationId, timeoutIdentifier);
+						}
+					});
+				}
+			}, slotRequestTimeout.toMilliseconds(), TimeUnit.MILLISECONDS);
 
-			LOG.info("Assigning SlotID({}) to AllocationID({}), JobID:{}", freeSlot.getSlotId(),
-				allocationId, chosenRequest.getJobId());
-			allocationMap.addAllocation(freeSlot.getSlotId(), allocationId);
+			pendingSlotRequest.registerTimeout(timeoutFuture, timeoutIdentifier);
 
-			sendSlotRequest(freeSlot, slotRequest);
-		} else {
-			freeSlots.put(freeSlot.getSlotId(), freeSlot);
+			resourceManagerActions.allocateResource(pendingSlotRequest.getResourceProfile());
 		}
 	}
 
-	private void sendSlotRequest(final ResourceSlot freeSlot, final SlotRequest slotRequest) {
+	/**
+	 * Allocates the given slot for the given slot request. This entails sending a registration
+	 * message to the task manager and treating failures.
+	 *
+	 * @param taskManagerSlot to allocate for the given slot request
+	 * @param pendingSlotRequest to allocate the given slot for
+	 */
+	private void allocateSlot(TaskManagerSlot taskManagerSlot, PendingSlotRequest pendingSlotRequest) {
+		TaskExecutorConnection taskExecutorConnection = taskManagerSlot.getTaskManagerConnection();
+		TaskExecutorGateway gateway = taskExecutorConnection.getTaskExecutorGateway();
+
+		final CompletableFuture<Acknowledge> completableFuture = new FlinkCompletableFuture<>();
+		final AllocationID allocationId = pendingSlotRequest.getAllocationId();
+		final SlotID slotId = taskManagerSlot.getSlotId();
+
+		taskManagerSlot.setAssignedSlotRequest(pendingSlotRequest);
+		pendingSlotRequest.setRequestFuture(completableFuture);
+
+		// RPC call to the task manager
+		Future<Acknowledge> requestFuture = gateway.requestSlot(
+			slotId,
+			pendingSlotRequest.getJobId(),
+			allocationId,
+			pendingSlotRequest.getTargetAddress(),
+			leaderId,
+			taskManagerRequestTimeout);
+
+		requestFuture.handle(new BiFunction<Acknowledge, Throwable, Void>() {
```

Comment:
The reason why we have this indirection is because we want to be able to cancel pending slot requests (e.g. if the SlotManager is suspended) from the `SlotManager` side.